Manwa: Optimize implementation logic

peakedshout opened the pull request:
1. Dynamically obtain the list of optional image sources;
2. Dynamically obtain the redirected access URL;
3. Reconstruct some access logic;
4. Support filtering and next page functions;

Closes #8999
Closes #8295

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

peakedshout committed:
Optimization:
1. Dynamically obtain the list of optional image sources;
2. Dynamically obtain the redirected access URL;
3. Reconstruct some access logic;
4. Support filtering and next page functions;

peakedshout committed:
Merge branch 'main' into f_manwa_v1.4.11

peakedshout committed:
Optimization:
1. Dynamically obtain the list of optional image sources;
2. Dynamically obtain the redirected access URL;
3. Reconstruct some access logic;
4. Support filtering and next page functions;

peakedshout committed:
Optimization:
1. Dynamically obtain the list of optional image sources;
2. Dynamically obtain the redirected access URL;
3. Reconstruct some access logic;
4. Support filtering and next page functions;

peakedshout committed:
Merge branch 'main' into f_manwa_v1.4.11

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/ImageSource.kt:
```diff
@@ -0,0 +1,83 @@
+
+                val responseBody = response.body.string()
+
+                val doc = Jsoup.parse(responseBody)
```
```suggestion
                val doc = response.asJsoup()
```

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/ImageSource.kt:
```diff
@@ -0,0 +1,83 @@
+                val responseBody = response.body.string()
+
+                val doc = Jsoup.parse(responseBody)
+                val modalBody = doc.select("#img-host-modal > div.modal-body").first()
```
```suggestion
                val modalBody = doc.selectFirst("#img-host-modal > div.modal-body")
```

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/Manwa.kt:
```diff
@@ -109,18 +130,99 @@ class Manwa : ParsedHttpSource(), ConfigurableSource {
+        }
+        if (page > 1) {
+            url = "$url&page=$page"
+        }
```
```suggestion
        val url  = baseUrl.toHttpUrl().newBuilder().apply {
            if (query != "" && !query.contains("-")) {
                addPathSegment("search")
                addQueryParameter("keyword", query)
            } else {
                (if (filters.isEmpty()) getFilterList() else filters).forEach { filter ->
                    when (filter) {
                        is UriPartFilter -> {
                            addQueryParameter(filter.toUriPart())
                        }
            
                        is TagCheckBoxFilterGroup -> {
                            addQueryParameter(filter.buildParams())
                        }
            
                        else -> {}
                    }
                }
            }
            if (page > 1) {
                addQueryParameter("page", page.toString())
            }
        }
```
may need to adjust lint

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/Manwa.kt:
```diff
@@ -109,18 +130,99 @@ class Manwa : ParsedHttpSource(), ConfigurableSource {
+        isUpdateTag = true
+        if (tagsJ != preferences.getString(APP_TAG_LIST_KEY, "")!!) {
+            preferences.edit().putString(APP_TAG_LIST_KEY, tagsJ).apply()
+        }
```
no need to compare, always just update

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/Manwa.kt:
```diff
@@ -153,72 +255,197 @@ class Manwa : ParsedHttpSource(), ConfigurableSource {
+        ),
+    )
+
+    private class EndFilter : UriPartFilter(
```
consider putting filters in separate file

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/Manwa.kt:
```diff
@@ -153,72 +255,197 @@ class Manwa : ParsedHttpSource(), ConfigurableSource {
+            TagCheckBoxFilter(k, v)
+        },
+    ) {
+        fun buildParams(): String {
```
adjust here for the searchrequest suggestion

peakedshout replied:
What mean?

AwkwardPeak7 replied:
referring to [this suggestion](https://github.com/keiyoushi/extensions-source/pull/9103#discussion_r2134349782)

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/Manwa.kt:
```diff
@@ -153,72 +255,197 @@ class Manwa : ParsedHttpSource(), ConfigurableSource {
+
+            entries = list
+            entryValues = list
+            setDefaultValue(list[0])
```
previously it saves the index, if you are changing the thing which is saved, either change the preference key or add a migration (to the getPreferences function) to migrate existing preferences

peakedshout replied:
I don't think the original key-values ​​are relevant now, even though the same placements are used.

AwkwardPeak7 replied:
it is not relevant to this code, but existing users will have them in the preference, which may interfere with this new logic

peakedshout replied:
I don't understand. The optional values ​​of the previous version are broken, and there is no point in migrating.
And the setting of selecting url should take the latest one to be correct.

AwkwardPeak7 replied:
if you think old preference value won't interfere/cause issues then that is fine

AwkwardPeak7 commented on src/zh/manwa/src/eu/kanade/tachiyomi/extension/zh/manwa/Manwa.kt:
```diff
@@ -153,72 +255,197 @@ class Manwa : ParsedHttpSource(), ConfigurableSource {
+                "https://manwass.cc/",
+                "https://manwatg.cc/",
+                "https://manwast.cc/",
+                "https://manwasy.cc/",
```
remove trailing slash

peakedshout replied:
The slash is retained here because the dynamically obtained URL (the response list) is also in this style. So I think there should be no problem with retaining the slash.

AwkwardPeak7 replied:
I would save trim those as well.
it's just that trailing slash can end up making urls in form: `https://www.example.com//some-thing` which sometimes end up as 404

peakedshout committed:
Optimization:
1. Dynamically obtain the list of optional image sources;
2. Dynamically obtain the redirected access URL;
3. Reconstruct some access logic;
4. Support filtering and next page functions;

peakedshout committed:
Merge branch 'main' into f_manwa_v1.4.11
